### ISSUE TITLE
Discussion for Azure Functions 3.0 Announcement

### ISSUE THREAD

**Issue by fabiocav** [Azure/app-service-announcements-discussions#115]
Discussion for issue https://github.com/Azure/app-service-announcements/issues/200

**Comment by PureKrome** [comment]
![](https://media.giphy.com/media/ckeHl52mNtoq87veET/giphy.gif)

![](https://media.giphy.com/media/vMEjhlxsBR7Fe/giphy.gif)

**Comment by MarkIannucci** [comment]
Any idea where we can find information on when Azure App Service will start working with .Net Core 3.0 without having to use the single file deployment workaround?  We're trying to make a decision on if we should use the single file workaround and knowing Microsoft's timeline would help us make that decision.

There are a few people raising the question on the [.Net Core 3.0 Availability blog post.](https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0/#comment-3257)

**Comment by mswietlicki** [comment]
@MarkIannucci You can always use "Web App for Containers" and host .Net Core 3.0 docker image on Azure App Service ;)

**Comment by mswietlicki** [comment]
By the way because of issue https://github.com/Azure/azure-functions-dotnet-extensions/issues/29 Azure Function 2.0 right now does not work with any Microsoft.Extensions.* 3.* packages and cannot share code with .Net Core 3.0 services.

So I hope Azure Function 3 will be available ASAP.

**Comment by jespersh** [comment]
> A public preview will be available on October 2019
> General Availability, running on .NET Core 3.1 (LTS release), scheduled for Q1 2020

Does this mean the entire 3.0 period will be preview until 3.0 GA in Q1?

**Comment by ArieJones** [comment]
Soooo.. any idea when updated nuget packages for a 3.0 compatible azure functions .. the 2.0 package issues detail above is definitely curtailing our efforts to update. Don't know about anyone else but business entities already seem to be freaking out unnecessarily maybe .. about this December 2.2 end of life and are pushing to get their environments updated.. 

So totally no pressure :)

**Comment by pseabury** [comment]
Like @ArieJones , the lack of ability to update our functions apps to Core3 packages is pretty rough.  The announcement says 

Timelines
A public preview will be available on October 2019
General Availability, running on .NET Core 3.1 (LTS release), scheduled for Q1 2020

Which includes the language "on October 2019" instead of "in October 2019" which indicates to me that a day was intended to be specified but wasn't.  Can we clarify is it closer to October 1 or October 30?

**Comment by SimonOrdo** [comment]
Any updates on the timeline?

**Comment by jeffhollan** [comment]
It's coming very soon - IN October, in the early part of October even I could share.  We're just working to make sure all the bits can be deployed and work before we release steps on how to use.

**Comment by pereiren** [comment]
Great news @jeffhollan Really hyped about this!

**Comment by pseabury** [comment]
I'm assuming you're more or less coordinating with https://github.com/Azure/app-service-announcements/issues/204 ?

> .NET Core 3.0 deployment on App Service Windows has started. We expect global deployment to complete by **October 11th.**
> 
> We'll update this announcement to notify of deployment completion or if the estimated completion date changes.

Thanks to all of you guys and your teams for this.  Really helps us as App Devs to be able to target a single runtime.

**Comment by electricessence** [comment]
Any update here?  Anyway we can get to tinker with the preview soon?

**Comment by sayanghosh123** [comment]
Would love to know when the preview is available ...

**Comment by SimonOrdo** [comment]
I really wish these updates were better coordinated between the different products.  If you have a multi-project solution, it's not reasonable to have different projects on different SDK versions.  Not to mention, the Azure environments for Web and Functions being out of sync and behind the SDK releases... it's a mess.

**Comment by fabiocav** [comment]
Apologies for the delayed response here.

We have just finalized validation and tooling changes and will begin the deployment process. We'll continue to provide updates in the announcement issue, but expect the preview to be available by 10/25. Thank you for the patience.

**Comment by SimonOrdo** [comment]
Ugh...

**Comment by fabiocav** [comment]
> Ugh...

@SimonOrdo , trying to better understand your reaction here. Is your dissatisfaction with the date mentioned above? If so, is there something you feel we could have made clearer in the announcement?

We did state this preview would be available in October, which aligns with the information I've provided above. 

Also, please note that, although we're committing to have that available **by** the 25th, our goal is to have that available earlier.

**Comment by andreasohlund** [comment]
Will we need to use a vs2019 preview once this happens? (I noticed that while most C# 8 things compile already using async streams doesn't )

**Comment by electricessence** [comment]
All I care about is that I'm not trapped in a maze of discontinued dependencies that prevent me from progressing.  If I can update my dependencies and still have the functions work, I'll be very happy.

**Comment by AbhiOnGithub** [comment]
So we will get Azure Function V3 preview soon , but what is the timelines that Azure Team is targeting to make it GA with V3.
Is it going to GA with dotnet core 3.0 or  will it be GA directly for dotnetcore 3.1

**Comment by pseabury** [comment]
Not only the crazy fixed dependencies (like json.net exact old version), but the fact that we have products that have a single/common codebase that use the .NET stack.  Desktop Apps, WebApps, WebApis, Functions that if we upgrade one piece to .NETCore3 we are stuck in a hard place with the rest.  For me currently, we had to create a new set of facade projects that uses links to the code in the .NET Core projects but targets Framework 4.x,  This is a pain, and would be worse if I was using the cloud build pipeline because the SDK versioning is also not synced with the other 3.0 bits.  Of course I could have moved nothing forward and not had this problem :-| 

I guess my feedback can best be summarized by saying - For some of your customers, it very much make a difference for the better when you move all of the now disconnected bits forward in a more synchronized manner.   I know some customers probably care nearly zero, but for those like us it would remove some pain even if you think it's self-inflicted.

**Comment by tomkerkhove** [comment]
Don't get me wrong guys but give the PG some slack - .NET Core 3.0 was just released so it's fair that it's still rolling out.

Better to use this approach than prematurely release .NET Core 3.0 support which will not work and cause outages.

Just my $0.02.

**Comment by SimonOrdo** [comment]
> > Ugh...
> 
> @SimonOrdo , trying to better understand your reaction here. Is your dissatisfaction with the date mentioned above? If so, is there something you feel we could have made clearer in the announcement?
> 
> We did state this preview would be available in October, which aligns with the information I've provided above.
> 
> Also, please note that, although we're committing to have that available **by** the 25th, our goal is to have that available earlier.

It's personal, i.e. specific to my situation.  We started the upgrades with the, perhaps outlandish, expectation that the platform bits would be available closer to the release of the SDK.  It's been a painful upgrade with tons of breaking changes and nebulously defined dates and now I have to go back to management and tell them it's yet another 10 days before we can test/deploy.  Like I said - a personal problem.

**Comment by ArieJones** [comment]
I think it would be a reasonable expectation that on announcing 3.0's release that all business ops within Microsoft would be in line for the release. So that a 3.0 release is truly a 3.0 release. Most of our projects are aligned like some of the previous commenters have detailed... with dependencies that with some of these changes we can't upgrade "just one" …. 
I think the majority of customers are probably looking at the timeline for 2.2 EOL coming up and that ramp gets shorter and shorter the longer they have to wait.. 

Just my 2 cents.

**Comment by fabiocav** [comment]
The preview bits are now available!

Please see this post for instructions on how to use the preview on Azure and how to the 3.0 tools locally:
https://dev.to/azure/develop-azure-functions-using-net-core-3-0-gcm

Please reach out with feedback and any potential issues you encounter and again, thank you for the patience!

**Comment by shibayan** [comment]
I tried Entity Framework Core 3 (netstandard2.1) with Azure Functions v3 (netcoreapp3.0), an error occurred in Metadata Generator.

```
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error : System.IO.FileNotFoundException: Could not load file or assembly 'netstandard, Version=2.1.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'. 指定されたファイルが見つかりません。
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error : File name: 'netstandard, Version=2.1.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51'
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error :    at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error :    at System.Reflection.RuntimeAssembly.GetExportedTypes()
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error :    at MakeFunctionJson.FunctionJsonConverter.TryGenerateFunctionJsons()
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error :    at MakeFunctionJson.FunctionJsonConverter.TryRun()
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error : 
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error : 
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error : Error generating functions metadata
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error :
Microsoft.NET.Sdk.Functions.Build.targets(41,5): error : Metadata generation failed.
```

#### Repro

1. Create new function v3 project and change TFM to `netcoreapp3.0`
2. Install Entity Framework Core 3 (`Microsoft.EntityFrameworkCore.SqlServer`)
3. Add `DbContext` implementation
4. Build

**Comment by jakubmaguza** [comment]
I'm currently targeting .net core 3 and azure functions ~2. It's working, so my question is:
Should I roll back to .net core 2, or .net core 3 is supported on azure functions ~2?

**Comment by andreasohlund** [comment]
Some things doesn't work on ~2 in my experience:

* System.Text.Json: blow up at runtime
* Async streams: won't compile

**Comment by Smartymind** [comment]
I tried to use SignalR Service with Azure Functions v3 but an error occured:

`[19/10/2019 14:00:07] Building host: startup suppressed:False, configuration suppressed: False
[19/10/2019 14:00:07] Reading functions metadata
[19/10/2019 14:00:07] 1 functions found
[19/10/2019 14:00:07] Loading startup extension 'SignalR'
[19/10/2019 14:00:07] Loaded extension 'SignalR' (1.0.1.0)
[19/10/2019 14:00:07] Loading startup extension 'AzureStorage'
[19/10/2019 14:00:07] Loaded extension 'AzureStorage' (3.0.10.0)
[19/10/2019 14:00:07] Initializing Warmup Extension.
[19/10/2019 14:00:08] A host error has occurred
[19/10/2019 14:00:08] Microsoft.Azure.SignalR.Common: Object reference not set to an instance of an object.
[19/10/2019 14:00:08] Initializing Warmup Extension.
Value cannot be null. (Parameter 'provider')
Press any to continue....[19/10/2019 14:00:09] Building host: startup suppressed:False, configuration suppressed: False`

Repro

1. Create new function v3 project and change TFM to netcoreapp3.0
2. Add SignalR NuGet package (Microsoft.Azure.WebJobs.Extensions.SignalRService)
3. Build

It does not even need to add any function parameter to reproduce the issue but only add SignalR Service extension to the packages reference.

By the way I have the same [issue](https://github.com/Azure/azure-functions-host/issues/5106) with EF Core 3 as @shibayan has experienced because we have refeence to a class library where the EF Core 3 is used.

**Comment by StefanSchoof** [comment]
Can you give any information when to expect the Coldstart optimisation in other regions? I am especially interested in West Europe.

And are there any other benefits then be able to use dotnet core 3? (No that this is not enough for me to get excited.)

**Comment by rybkov** [comment]
When trying to migrate V2 functin to V3 getting this compilation error

System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.AspNetCore.Mvc.Abstractions, Version=3.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The system cannot find the file specified.

It happens when reference project that is Microsoft.AspNetCore.Http dependency

**Comment by electricessence** [comment]
Just reporting in:
```
Severity  Code  Description  Project  File  Line  Suppression State
Error    System.IO.FileNotFoundException: Could not load file or assembly 'System.Threading.Tasks.Dataflow, Version=4.6.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
File name: 'System.Threading.Tasks.Dataflow, Version=4.6.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
   at System.Reflection.RuntimeAssembly.GetExportedTypes(RuntimeAssembly assembly, ObjectHandleOnStack retTypes)
   at System.Reflection.RuntimeAssembly.GetExportedTypes()
   at MakeFunctionJson.FunctionJsonConverter.TryGenerateFunctionJsons()
   at MakeFunctionJson.FunctionJsonConverter.TryRun()


Error generating functions metadata
  Payscale.Consumer.Functions.RC.Queuing  C:\Users\[User]\.nuget\packages\microsoft.net.sdk.functions\1.0.30-beta2\build\Microsoft.NET.Sdk.Functions.Build.targets  41
```

**Comment by electricessence** [comment]
@fabiocav Can you clarify something for me?
AFv3 is backwards compatible with AFv2.  Got that.
But does AFv3 work with .NET Core 2.2?  Or only .NET Core 3?
And is it true that AFv2 still works under .NET Core 3?

**Comment by mswilson4040** [comment]
Anyone hit this error yet?

> Microsoft.Data.SqlClient: Microsoft.Data.SqlClient is not supported on this platform.

https://github.com/Azure/Azure-Functions/issues/1370

**Comment by paulbatum** [comment]
Thanks for the issue reports folks! We have a couple of folks out of the office right now so our responses on some of these reports might be a bit delayed. A few quick answers:

@StefanSchoof cold start optimization for all regions will slowly roll out over the next month.

@electricessence I tried to repro your issue by just adding a reference to the latest version of System.Thread.Tasks.Dataflow, but it worked fine. One thing I noticed is that your error message refers to 4.6.4, but webjobs has a dependency on 4.8.0, so I would suggest trying to update. Also double check that Microsoft.NET.Sdk.Functions is at 1.0.30-beta2. If none of that helps, please file a new issue and leave a link here.

@electricessence Functions V3 runs on .NET Core 3.0. When you say "does it work with .NET Core 2.2", I assume you mean "is it OK if my V3 functions are in a class library that targets 2.2?" the answer is that there is a good chance it will continue working, but you're better off updating your functions class libraries projects to target .NET Core 3.0 to minimize the chances of hitting differences between 2.2 and 3.0. For your last question - Functions V2 currently runs on .NET Core 2.2, not sure what else you were asking for?

@mswilson4040 I can confirm that its easy to repro the Microsoft.Data.SqlClient issue. Doesn't look V3 specific either. Since there are few issues filed already, conversation can continue there.

**Comment by jakubmaguza** [comment]
Thanks for answers. 
What about targeting class library for .net core 3.0, and using azure functions ~2? It's working now. Is it because I'm not using anything 3.0 specific?

**Comment by paulbatum** [comment]
@jakubmaguza Yes correct. For example, try adding the following line of code to your function: 

```csharp
Index i = 1;
```
This will compile (since you are targeting .NET Core 3.0), but when you run on functions V2 you'll run on .NET Core 2.2 and it will blow up with this error:

```
[10/23/2019 10:19:12 PM] The following 1 functions are in error:
[10/23/2019 10:19:12 PM] Function1: Unable to load one or more of the requested types.
[10/23/2019 10:19:12 PM] Could not load type 'System.Index' from assembly 'System.Runtime, Version=4.2.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'.
```

The same line of code works fine on Functions V3.

**Comment by electricessence** [comment]
@paulbatum Yeah, I'm baffled because the actual reference to Dataflow imported is 4.9 :|
I don't think this specific error occurs with .NET Core 2.2?  I kinda gave up for now. :(

**Comment by SimonOrdo** [comment]
Should `AddHttpClient()` work with V3 assemblies?  I've attempted to set it up as per the preview release instructions, but my V2 project had a call to `builder.Services.AddHttpClient()`.  As such:

```
using System;
using Data;
using Microsoft.Azure.Functions.Extensions.DependencyInjection;
using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Http; // <-- Not found
using Microsoft.Extensions.Logging;
```

```
  public class Startup : FunctionsStartup
    {
        public IConfiguration Configuration { get; set; }

      public override void Configure(IFunctionsHostBuilder builder)
        {
            builder.Services.AddOptions()
                .Configure<IConfiguration>((configuration) =>
                {
                    Configuration = configuration;
                });

            builder.Services.AddHttpClient();
        }
    }
```

which gives the error :

> The type or namespace name 'Http' does not exist in the namespace 'Microsoft.Extensions' (are you missing an assembly reference?)`

Project appears as follows:

```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.Functions.Extensions" Version="1.0.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.30-beta2" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\DataAccess\DataAccess.csproj" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

**Comment by x7even** [comment]
> All I care about is that I'm not trapped in a maze of discontinued dependencies that prevent me from progressing. If I can update my dependencies and still have the functions work, I'll be very happy.

This is the underrated comment.

**Comment by IanKemp** [comment]
Is the [SDK](https://www.nuget.org/packages/Microsoft.NET.Sdk.Functions) still going to be confusingly-versioned at 1.x, or are you actually going to bump the major version so it matches the runtime version?

**Comment by gajduk-mansystems** [comment]
I am running a Node.js function. What does this mean for people like me?

**Comment by paulbatum** [comment]
@IanKemp yes we are looking at bumping the version this time. We need to be careful about avoiding a situation where V1/V2 users go into nuget and click "update all" and are broken. We have some ideas about how to make sure this doesnt happen, just working on the details.

**Comment by paulbatum** [comment]
@gajduk-mansystems For V3 we have a small set of JavaScript bugfixes planned that are technically breaking changes. Its also likely there will be a change to the supported node versions. We will share another update with these details.

**Comment by biltongza** [comment]
Posting here for awareness since it's been nearly a month with no official response: https://github.com/Azure/Azure-Functions/issues/1380

**Comment by shibayan** [comment]
The v3 runtime seems to have disappeared from the feed. I wanted to try the 3.0 SDK, but I couldn't create a v3 project.

Rel : https://github.com/Azure/azure-functions-tooling-feed/issues/131#issue-521917990

**Comment by pereiren** [comment]
Hi @shibanik , I had same issue 2 weeks ago or so and I found this workaround here: https://dev.to/sjscheidertt/comment/ho83

Hope it helps

**Comment by hiraldesai** [comment]
I have upgraded one of our apps to v3 preview and everything is working as expected thanks to the [great article](https://dev.to/azure/develop-azure-functions-using-net-core-3-0-gcm). Question: it is recommended not to go to production with this (it is tempting though!), so when will this be out of preview so I can make the necessary changes and deploy to production?

**Comment by jeffhollan** [comment]
@hiraldesai very soon - our goal is to have a production ready release potentially as early as next week. Stay tuned to the app-service-announcement repo for updates and will announce as soon as that lands

**Comment by fabiocav** [comment]
To add to what @jeffhollan mentioned, the next drop of 3.0 will target and support .NET Core 3.1

**Comment by fabiocav** [comment]
The latest Azure Functions 3.0 release is now globally deployed and is a Go Live release.

For more information, questions or comments, please see: https://github.com/Azure/app-service-announcements/issues/220

**Comment by fabiocav** [comment]
Thank all you for the feedback and help during the preview period!

**Comment by electricessence** [comment]
Finally upgraded and got things working.  Thanks guys!
I still wish it was possible to use the more updated Azure Storage libs (Microsoft.Azure instead of Microsoft.WindowsAzure), but I'm content for now. :)

**Comment by paulbatum** [comment]
@electricessence thats great to hear. We have an update close to being ready for storage v11 (Microsoft.Azure) which you can track [here](https://github.com/Azure/azure-webjobs-sdk/pull/2356).
